dvc/objects/stage.py

```diff
@@ -174,11 +172,7 @@ def stage(odb, path_info, fs, name, upload=False, **kwargs):
         isinstance(path_info, str) or path_info.scheme == fs.scheme
     )
 
-    if not fs.exists(path_info):
-        raise FileNotFoundError(
-            errno.ENOENT, os.strerror(errno.ENOENT), path_info
-        )
-
+    details = fs.info(path_info)
```

Comment:
Ideally, we shouldn't need to call this here at all, but that requires logic below to properly raise FileNotFound (e.g. `state.get()` in case when it is not a `StateNoop`, and `_get_file_obj` if it is a Noop). I'm almost sure that should work already, but need to check that. Could you please give it a look?

Reply:
We still need to make an info call, but just later in that case. Not sure what would be changed? Also isn't for remote providers (e.g s3) the state is `StateNoop`?

Reply:
@isidentical We would live `isdir()` as it was before. In current PR we are calling info() once in place of old exits(), but then only using it instead of `isdir`.  So my point is rather about `fs.exists()` not being needed right at that place, so `info()` call is also not needed there.

> Also isn't for remote providers (e.g s3) the state is StateNoop?

Correct, which means that in current form `state.get()`'s behavior will differ between `State` and `StateNoop`, which might be problematic. But leaving `info()` where it is right now, makes that part of the code the slightest bit more confusing than it was before (at least it was an explicit exists() call).

Thinking about it a bit more, it becomes clear to me that this is probably a topic for a followup, as that `state.get` call could also be moved a bit. Anyway, let's merge and we'll get back to the follups in the future.

Reply:
> Correct, which means that in current form state.get()'s behavior will differ between State and StateNoop, which might be problematic. But leaving info() where it is right now, makes that part of the code the slightest bit more confusing than it was before (at least it was an explicit exists() call).

If we defer this check to `state.get()`, then for example when you are staging something over s3 and if the directory doesn't exist yet this might cause a different behavior. So that was my worry about not shifting this and depending solely on `state.get`

Reply:
@isidentical Yeah, makes sense! That's what I meant in my last comment too, we'll likely move that `state.get` in the future and will get back to this particular place in the code.